File: Part 2 - Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.py
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv("50_Startups.csv")
X = dataset.iloc[:, :-1].values
Y = dataset.iloc[:, -1].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,-1] = labelencoder_X.fit_transform(X[:,-1])
onehotencoder = OneHotEncoder(categorical_features=[-1])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
# remove uma das colunas das dummy variables
# a bibliotega de regressão linear do python já faz isso, mas algumas bibliotecas não fazem, então é sempre bom fazer
X = X[:,1:]

# Splitting the dataset in train set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

# Feature scaling
# a biblioteca aplica feature scalling internamente automaticamente
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
# IMPORTANTE: aqui faz só o transform, pois o fit já foi feito no train, isso faz com que ambos esteja escaladas na mesma base
X_test = sc_X.transform(X_test)
# em classificações o Y não precisa ser escalado. Em regreção sim
"""

# Multiple Linear Regression

# Fitting Multiple Linear Regression to the Training set
# no fit o algoritmo estuda e aprende a corelação entre as variáveis independentes e a dependente
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
reg.fit(X_train, Y_train)

# Predicting the Test set results
Y_pred = reg.predict(X_test)

# Building the optimal model using Backward Elimination
# adiciona a coluna x0 = 1 para o algoritmo reconehcer o b0 da equação, isso é considerado automaticamente no LinearRegression, mas aqui não. 
# isso é avisado no ctrl+i do OLS
import statsmodels.api as sm
X = np.append(arr = np.ones(shape = (50,1)).astype(int), values = X, axis = 1)
X_opt = X[:,[0,1,2,3,4,5]]
reg_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
reg_OLS.summary()
# remove colunas enquanto o maior p for maior que o limite selecionado, nesse exemplo 0.05, ou seja, 5%
# remove x2
X_opt = X[:,[0,1,3,4,5]]
reg_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
reg_OLS.summary()
# remove x1
X_opt = X[:,[0,3,4,5]]
reg_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
reg_OLS.summary()
# remove x2
X_opt = X[:,[0,3,5]]
reg_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
reg_OLS.summary()
# remove x2
X_opt = X[:,[0,3]]
reg_OLS = sm.OLS(endog = Y, exog = X_opt).fit()
reg_OLS.summary()

# Predicting the optimal Test set results
X_train_2, X_test_2, Y_train_2, Y_test_2 = train_test_split(X_opt, Y, test_size = 0.2, random_state = 0)
reg_2 = LinearRegression()
reg_2.fit(X_train_2, Y_train_2)
Y_opt_pred = reg_2.predict(X_test_2)

File: Part 2 - Regression/Section 5 - Multiple Linear Regression/multiple_linear_regression.R
# Importing the dataset
dataset = read.csv("50_Startups.csv")

# Encoding categorical data
# com esse processo a coluna assume valores do tipo factor, e não númericos, apesar de serem números
# tem que manter como factor para transformar em dummy variable e trata o dummy variale trap, ambos automaticamente 
dataset$State = factor(dataset$State,
                         levels = c("New York","California","Florida"),
                         labels = c(1,2,3))

# Splitting the dataset in train set and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
train_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature scaling
# a biblioteca aplica feature scalling internamente, assim como no python
# train_set$Country = as.numeric(train_set$Country)
# test_set$Country = as.numeric(test_set$Country)
# train_set[,1:3] = scale(train_set[,1:3])
# test_set[,1:3]  = scale(test_set[,1:3])

# Multiple Linear Regression

# Fitting Multiple Linear Regression to the Training set
# no fit o algoritmo estuda e aprende a corelação entre as variáveis independentes e a dependente
# a formula recebe como paremetros a correlção entre as colunas do data
#reg = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
#         data = train_set)
reg = lm(formula = Profit ~ .,
         data = train_set)
# summary(reg)

# Predicting the Test set results
# tranformei em dataframe somente para visualizar no Environment
Y_pred = as.data.frame(predict(reg, newdata = test_set))

# Predicting the Test set results with the best predictor
# ambas não deram o mesmo resultado como dito no vídeo
reg = lm(formula = Profit ~ R.D.Spend ,
         data = train_set)
# summary(reg)
Y_pred_best_predictor = as.data.frame(predict(reg, newdata = test_set))

# Building the optimal model using Backward Elimination
# para fazer Backward Elimination se usa a mesma função, porem com um diferente sistema de parâmetros
reg = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
         data = dataset)
summary(reg)
# remove colunas enquanto o maior p for maior que o limite selecionado, nesse exemplo 0.05, ou seja, 5%
# remove State
reg = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
         data = dataset)
summary(reg)
# remove Administration
reg = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
         data = dataset)
summary(reg)
# remove Marketing.Spend
reg = lm(formula = Profit ~ R.D.Spend,
         data = dataset)
summary(reg)

# Predicting the optimal Test set results
reg = lm(formula = Profit ~ R.D.Spend ,
         data = train_set)
Y_opt_pred = as.data.frame(predict(reg, newdata = test_set))
